shell/platform/linux/fl_renderer_wayland.cc

```diff
@@ -0,0 +1,322 @@
+                "Expected Wayland display");
+    return FALSE;
+  }
+  GdkWaylandDisplay* gdk_display =
+      GDK_WAYLAND_DISPLAY(gdk_display_get_default());
+
+  wl_compositor* compositor =
+      gdk_wayland_display_get_wl_compositor(gdk_display);
+  if (!compositor) {
+    g_set_error(error, fl_renderer_error_quark(), FL_RENDERER_ERROR_FAILED,
+                "No wl_compositor");
+    return FALSE;
+  }
+
+  // Make sure size and scale is not <= 0
+  self->subsurface.scale = MAX(self->subsurface.scale, 1);
+  self->subsurface.geometry.width = MAX(self->subsurface.geometry.width, 1);
+  self->subsurface.geometry.height = MAX(self->subsurface.geometry.height, 1);
+
+  self->subsurface.surface = wl_compositor_create_surface(compositor);
+  self->resource.surface = wl_compositor_create_surface(compositor);
+  if (!self->subsurface.surface || !self->resource.surface) {
+    g_set_error(error, fl_renderer_error_quark(), FL_RENDERER_ERROR_FAILED,
+                "Failed to create wl_surfaces");
+    return FALSE;
+  }
+  wl_surface_set_buffer_scale(self->subsurface.surface, self->subsurface.scale);
+
+  gint window_width = self->subsurface.geometry.width * self->subsurface.scale;
+  gint window_height =
+      self->subsurface.geometry.height * self->subsurface.scale;
+  self->subsurface.egl_window = wl_egl_window_create(
+      self->subsurface.surface, window_width, window_height);
+  self->resource.egl_window =
+      wl_egl_window_create(self->resource.surface, 1, 1);
+  if (!self->subsurface.egl_window || !self->resource.egl_window) {
+    g_set_error(error, fl_renderer_error_quark(), FL_RENDERER_ERROR_FAILED,
+                "Failed to create EGL windows");
+    return FALSE;
+  }
+
+  *visible = eglCreateWindowSurface(egl_display, config,
+                                    self->subsurface.egl_window, nullptr);
+  *resource = eglCreateWindowSurface(egl_display, config,
+                                     self->resource.egl_window, nullptr);
+  if (*visible == EGL_NO_SURFACE || *resource == EGL_NO_SURFACE) {
+    EGLint egl_error = eglGetError();  // must be before egl_config_to_string()
+    g_autofree gchar* config_string = egl_config_to_string(egl_display, config);
+    g_set_error(error, fl_renderer_error_quark(), FL_RENDERER_ERROR_FAILED,
+                "Failed to create EGL surfaces using configuration (%s): %s",
+                config_string, egl_error_to_string(egl_error));
+    return FALSE;
+  }
+
+  GtkWidget* toplevel = fl_renderer_wayland_get_toplevel_widget(self);
+  if (!toplevel) {
+    g_set_error(error, fl_renderer_error_quark(), FL_RENDERER_ERROR_FAILED,
+                "Renderer does not have a widget");
+    return FALSE;
+  }
+  g_signal_connect(toplevel, "map",
+                   G_CALLBACK(fl_renderer_wayland_on_window_map), self);
+  g_signal_connect(toplevel, "unmap",
+                   G_CALLBACK(fl_renderer_wayland_on_window_unmap), self);
```

Comment:
Use `g_signal_connect_object` here, then the `g_signal_handlers_disconnect_by_func` calls can be removed from `fl_renderer_wayland_dispose`.

Reply:
And the function prototypes can be removed from the top of this file.

Reply:
And please connect the signals `G_CONNECT_SWAPPED` (so they look like standard methods).

Reply:
Done